src/foam/nanos/ruler/RulerDAO.js

```diff
@@ -123,14 +172,93 @@
         }
       ],
       javaCode: `
-
       for ( Object key : sink.getGroupKeys() ) {
         List<Rule> group = ((ArraySink) sink.getGroups().get(key)).getArray();
         if ( ! group.isEmpty() ) {
+          Collections.sort(group, new foam.mlang.order.Desc(Rule.PRIORITY));
           new RuleEngine(x, this).execute(group, obj, oldObj);
         }
       }
       `
+    },
+    {
+      name: 'updateRules',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        }
+      ],
+      javaCode: `DAO ruleDAO = ((DAO) x.get("ruleDAO")).where(EQ(Rule.DAO_KEY, getDaoKey()));
+Map rulesList = getRulesList();
+GroupBy createdBefore = (GroupBy) ruleDAO.where(getCreateBefore()).select(GROUP_BY(Rule.RULE_GROUP, new ArraySink()));
+rulesList.put(getCreateBefore(), createdBefore);
+GroupBy updatedBefore = (GroupBy) ruleDAO.where(getUpdateBefore()).select(GROUP_BY(Rule.RULE_GROUP, new ArraySink()));
+rulesList.put(getUpdateBefore(), updatedBefore);
+GroupBy createdAfter = (GroupBy) ruleDAO.where(getCreateAfter()).select(GROUP_BY(Rule.RULE_GROUP, new ArraySink()));
+rulesList.put(getCreateAfter(), createdAfter);
+GroupBy updatedAfter = (GroupBy) ruleDAO.where(getUpdateAfter()).select(GROUP_BY(Rule.RULE_GROUP, new ArraySink()));
+rulesList.put(getUpdateAfter(), updatedAfter);
+GroupBy removedBefore = (GroupBy) ruleDAO.where(getRemoveBefore()).select(GROUP_BY(Rule.RULE_GROUP, new ArraySink()));
+rulesList.put(getRemoveBefore(), removedBefore);
+GroupBy removedAfter = (GroupBy) ruleDAO.where(getRemoveAfter()).select(GROUP_BY(Rule.RULE_GROUP, new ArraySink()));
+rulesList.put(getRemoveAfter(), removedAfter);
+
+ruleDAO.listen(new AbstractSink() {
+  @Override
+  public void put(Object obj, Detachable sub) {
+    Map rulesList = getRulesList();
+    Rule rule = (Rule) obj;
+    String ruleGroup = rule.getRuleGroup();
+    for ( Object key : rulesList.keySet() ) {
```

Comment:
One optimization we can consider is to check Rule.daoKey == RulerDAO.daoKey before running the loop. @penzital ?

Reply:
it's actually not for optimization purposes but an actual bug as I'm adding all the rules to the lists not only the ones that match daoKey which is wrong